### PR TITLE
log timestamps

### LINKED ISSUE #444: Improve logging
We could think about wrapping os.Stdout in the default logger, or using a log.Logger.

### PR DESCRIPTION
**Pull request by marten-seemann**
Print timestamps in log files.

Fixes #444.

### CODE REVIEW

**Comment by marten-seemann** [comment]
I'd like to keep a way to set the log level manually. I'm not really worried about the API, since utils is not part of our public API anyway (to make that even clearer, we can think about making it an internal package some time).

**Comment by lucas-clemente** [comment]
I'm fine with keeping utils.SetLogLevel. What do you think about using Go's log package?

**Comment by marten-seemann** [comment]
tbh, I haven't used it so far.
From the docs I'd say that we'd only use `Logger.SetPrefix` (for setting the timestamp prefix), `Logger.SetOutput` (to set the writer) and `Logger.Printf`, right?

**Comment by lucas-clemente** [comment]
Yes. Wrt the prefix or output, it actually has a proper defaults, which we definitely shouldn't change in non-test code. For testing code we can decide about the format, but I'd argue for keeping the default there as well.

**Comment by marten-seemann** [comment]
The `log` package is pretty unflexible. I just tried to use it.
Disabling timestamp works with the self-explanatory command `log.SetFlags(0)`. `log.SetPrefix` is pretty much useless, since it only takes a string, not a callback. And logging with millisecond resolution is impossible. I'd prefer to not use this package.

**Comment by lucas-clemente** [comment]
I don't think it's the job of a QUIC implementation to solve the logging problem. Especially if we ever want to get in there, we don't have a choice here. If this package works for the rest of the world, it should work for us.

Regarding the resolution, have you seen `log.Lmicroseconds`?

**Comment by marten-seemann** [comment]
Well, there is a reason there is a whole bunch of logging libraries around. Logging is definitely our smallest problem regarding standard library compatibility, so I don't see any need to move to an inconvenient solution because of that, at this point.
Since we'll use logging a lot for debugging, I prefer to use our own solution that provides an API that I don't have to look up in the godoc every single time.

Regarding the resolution: This is just a minor point, but yet another example of the bad API of `log`. I'm aware of `log.Lmicrosecond`, but there's no `log.Lmillisecond`, and no way to specify our own format.

**Comment by marten-seemann** [comment]
What we can do is keep the `SetLogTimeFormat` function. Then we can still use the `log` package, but we're completely free to do what we want for our tests.

**Comment by lucas-clemente** [comment]
I think the additional code is not worth saving 3 chars per log line. I'd strongly prefer to use Lmicroseconds and not have any custom code for the format / prefix.

**Comment by marten-seemann** [comment]
I pushed a commit that now uses the standard library log package. I think this is a good compromise fulfilling the logging needs of quic-go users on the one hand, while at the same time providing a more convenient API for internal purposes.

File: utils/log.go [hunk 1]
```diff
@@ -26,16 +25,13 @@ const (
...
-	out        io.Writer = os.Stdout
-	timeFormat           = "15:04:05.000"
-
-	mutex sync.Mutex
+	logLevel   = LogLevelNothing
+	timeFormat = ""
 )
 
 // SetLogWriter sets the log writer.
 func SetLogWriter(w io.Writer) {
```
Comment (lucas-clemente) [inline utils/log.go]: Remove this function, now that it doesn't do anything.

File: integrationtests/integrationtests_suite_test.go [hunk 2]
```diff
@@ -78,6 +78,9 @@ func init() {
 }
 
 var _ = BeforeEach(func() {
+	// set custom time format for logs
+	utils.SetLogTimeFormat("15:04:05.000")
```
Comment (lucas-clemente) [inline integrationtests/integrationtests_suite_test.go]: I still think it's kind of ridiculous have code in a QUIC library just for this, but if you really must…

File: utils/log.go [hunk 3]
```diff
@@ -46,42 +42,36 @@ func SetLogLevel(level LogLevel) {
...
-		mutex.Unlock()
+		logMessage(format, args...)
 	}
 }
 
-func logTimestamp() {
+func logMessage(format string, args ...interface{}) {
 	if len(timeFormat) > 0 {
-		fmt.Fprintf(out, time.Now().Format(timeFormat)+" ")
+		log.Printf(time.Now().Format(timeFormat)+" "+format, args...)
```
Comment (lucas-clemente) [inline utils/log.go]: This is a red flag from a security point of view (never have dynamic format strings). It doesn't really make a difference right now, but please do this in two prints.

Comment (marten-seemann) [inline utils/log.go]: That's not possible. `log.Printf` adds a newline.

**Comment by marten-seemann** [comment]
@lucas-clemente: lgty?

**Pull request merged**